tune/protox/agent/replay.py

```diff
@@ -0,0 +1,337 @@
+                assert original_workload_runtime > 0
+
+                # Extract the necessary values from action.pkl
+                with open_and_save(dbgym_cfg, tuning_steps_dpath / repo / "action.pkl", "rb") as f:
+                    actions_info = pickle.load(f)
+                    all_holon_action_variations = actions_info["all_holon_action_variations"]
+                    # Extract the KnobSpaceAction and IndexAction from all_holon_action_variations.
+                    # These two should be identical across all HolonActions, which we will assert.
+                    _, first_holon_action = all_holon_action_variations[0]
+                    knob_space_action = first_holon_action[0]
+                    index_space_raw_sample = first_holon_action[1]
+                    index_action = action_space.get_index_space().to_action(index_space_raw_sample)
+                    assert all([knob_space_action == holon_action[0] for (_, holon_action) in all_holon_action_variations])
+                    assert all([index_action == action_space.get_index_space().to_action(holon_action[1]) for (_, holon_action) in all_holon_action_variations])
+
+                # Get the indexes from this action and the prior state
+                index_acts = set()
+                index_acts.add(index_action)
+                assert len(index_acts) > 0
+                with open_and_save(dbgym_cfg, tuning_steps_dpath / repo / "prior_state.pkl", "rb") as f:
+                    prior_states = pickle.load(f)
+                    all_sc = set(prior_states[1])
+                    for index_act in index_acts:
+                        all_sc.add(index_act)
+
+                    all_sc = {a for a in all_sc if not "USING btree ()" in a.sql(True)}
+                    index_acts = all_sc
+                # Get the CREATE INDEX or DROP INDEX statements to turn the state into the one we should be in at this tuning step
+                index_modification_sqls = []
+                for index_act in index_acts:
+                    if index_act not in existing_index_acts:
+                        index_modification_sqls.append(index_act.sql(True))
+                for existing_index_act in existing_index_acts:
+                    if existing_index_act not in index_acts:
+                        index_modification_sqls.append(existing_index_act.sql(False))
+
+                # Modify Postgres to have the right indexes and system-wide knobs. `index_modification_sqls` holds the indexes
+                #   while `cc` holds the system-wide knobs.
+                if not replay_args.simulated:
+                    cc, _ = action_space.get_knob_space().generate_action_plan(knob_space_action, prior_states[0])
+                    # Like in tuning, we don't dump the page cache when calling shift_state() to see how the workload
+                    #   performs in a warm cache scenario.
+                    pg_env.shift_state(cc, index_modification_sqls, dump_page_cache=False)
```

Comment:
This seems to contradict what we actually do during tuning? Since we seem to be dumping the page cache there?